src/kubernetes.cc

```diff
@@ -794,33 +794,50 @@ void KubernetesReader::WatchMaster(
       "Authorization", "Bearer " + KubernetesApiToken());
   const bool verbose = config_.VerboseLogging();
   if (verbose) {
-    LOG(INFO) << "WatchMaster(" << name << "): Contacting " << endpoint;
+    LOG(INFO) << "WatchMaster(" << name << "): Will retry "
+              << config_.KubernetesUpdaterWatchRetries() << " times";
   }
-  try {
-    if (verbose) {
-      LOG(INFO) << "Locking completion mutex";
-    }
-    // A notification for watch completion.
-    std::mutex completion_mutex;
-    std::unique_lock<std::mutex> watch_completion(completion_mutex);
-    Watcher watcher(
-      endpoint,
-      [=](json::value raw_watch) {
-        WatchEventCallback(callback, name, std::move(raw_watch));
-      },
-      std::move(watch_completion), verbose);
-    http::client::response response = client.get(request, std::ref(watcher));
-    if (verbose) {
-      LOG(INFO) << "Waiting for completion";
-    }
-    std::lock_guard<std::mutex> await_completion(completion_mutex);
-    if (verbose) {
-      LOG(INFO) << "WatchMaster(" << name << ") completed " << body(response);
+  int failures = 0;
+  std::string last_error;
+  while (failures < config_.KubernetesUpdaterWatchRetries()) {
+    try {
+      if (verbose) {
+        LOG(INFO) << "WatchMaster(" << name << "): Contacting " << endpoint
+                  << " after " << failures << " failures";
```

Comment:
I would avoid logging this unless `failures` is greater than 0... Your earlier code that said "Retrying; attempt #..." was fine.